File: packages/alarm/ios/Plugin/AlarmPlugin.swift
import Foundation
import Capacitor

@objc(AlarmPlugin)
public class AlarmPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AlarmPlugin"
    public let jsName = "Alarm"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "cancelAlarm", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createAlarm", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createTimer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAlarms", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAlarms", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "AlarmPlugin"

    private var implementation: Any?

    override public func load() {
        if #available(iOS 26.0, *) {
            self.implementation = Alarm(plugin: self)
        }
    }

    @objc func cancelAlarm(_ call: CAPPluginCall) {
        guard #available(iOS 26.0, *), let implementation = implementation as? Alarm else {
            rejectCallAsUnavailable(call)
            return
        }
        do {
            let options = try CancelAlarmOptions(call)
            implementation.cancelAlarm(options, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call)
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        guard #available(iOS 26.0, *), let implementation = implementation as? Alarm else {
            rejectCallAsUnavailable(call)
            return
        }
        implementation.checkPermissions(completion: { result in
            self.resolveCall(call, result)
        })
    }

    @objc func createAlarm(_ call: CAPPluginCall) {
        guard #available(iOS 26.0, *), let implementation = implementation as? Alarm else {
            rejectCallAsUnavailable(call)
            return
        }
        do {
            let options = try CreateAlarmOptions(call)
            implementation.createAlarm(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call, result)
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func createTimer(_ call: CAPPluginCall) {
        rejectCallAsUnimplemented(call)
    }

    @objc func getAlarms(_ call: CAPPluginCall) {
        guard #available(iOS 26.0, *), let implementation = implementation as? Alarm else {
            rejectCallAsUnavailable(call)
            return
        }
        implementation.getAlarms(completion: { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        })
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        if #available(iOS 26.0, *), let implementation = implementation as? Alarm {
            implementation.isAvailable(completion: { result in
                self.resolveCall(call, result)
            })
        } else {
            resolveCall(call, IsAvailableResult(available: false))
        }
    }

    @objc func openAlarms(_ call: CAPPluginCall) {
        rejectCallAsUnimplemented(call)
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        guard #available(iOS 26.0, *), let implementation = implementation as? Alarm else {
            rejectCallAsUnavailable(call)
            return
        }
        implementation.requestPermissions(completion: { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        })
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", AlarmPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func rejectCallAsUnavailable(_ call: CAPPluginCall) {
        call.unavailable("This method is not available on this platform.")
    }

    private func rejectCallAsUnimplemented(_ call: CAPPluginCall) {
        call.unimplemented("This method is not available on this platform.")
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/alarm/android/src/main/java/io/capawesome/capacitorjs/plugins/alarm/AlarmPlugin.java
package io.capawesome.capacitorjs.plugins.alarm;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import io.capawesome.capacitorjs.plugins.alarm.classes.CustomException;
import io.capawesome.capacitorjs.plugins.alarm.classes.options.CreateAlarmOptions;
import io.capawesome.capacitorjs.plugins.alarm.classes.options.CreateTimerOptions;
import io.capawesome.capacitorjs.plugins.alarm.classes.results.CreateAlarmResult;
import io.capawesome.capacitorjs.plugins.alarm.classes.results.IsAvailableResult;
import io.capawesome.capacitorjs.plugins.alarm.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.alarm.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.alarm.interfaces.Result;

@CapacitorPlugin(name = "Alarm", permissions = @Permission(strings = {}, alias = "alarms"))
public class AlarmPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "AlarmPlugin";

    private Alarm implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new Alarm(this);
    }

    @PluginMethod
    public void cancelAlarm(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void createAlarm(PluginCall call) {
        try {
            CreateAlarmOptions options = new CreateAlarmOptions(call);
            NonEmptyResultCallback<CreateAlarmResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull CreateAlarmResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.createAlarm(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void createTimer(PluginCall call) {
        try {
            CreateTimerOptions options = new CreateTimerOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.createTimer(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getAlarms(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void isAvailable(PluginCall call) {
        try {
            NonEmptyResultCallback<IsAvailableResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull IsAvailableResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.isAvailable(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void openAlarms(PluginCall call) {
        try {
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.openAlarms(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void rejectCallAsUnimplemented(@NonNull PluginCall call) {
        call.unimplemented("This method is not available on this platform.");
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
